### PR TITLE
bfd, copp and dhcp relay and snooping check and diff modes

### PR DESCRIPTION
**Pull request by mingjunzhang2019**
##### SUMMARY
Th is to add diff and check modes support for copp, bfd, dhcp_relay and dhcp_snooping modules.

##### GitHub Issues
N/A

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
sonic_bfd, sonic_copp, sonic_dhcp_relay, sonic_dhcp_snooping

##### OUTPUT
- bfd_regression result: 
[bfd_regression-2024-02-23-00-22-13.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409158/bfd_regression-2024-02-23-00-22-13.html.pdf)
- bfd regression log: 
[bfd_regression.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409160/bfd_regression.log)
- bfd unit test log: 
[bfd_unit_test.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409162/bfd_unit_test.log)
- bfd unit test script: 
[bfd_unit_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409167/bfd_unit_test.txt)

- copp regression result: 
[copp_regression-2024-02-21-18-21-23.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409172/copp_regression-2024-02-21-18-21-23.html.pdf)
- copp regression log: 
[copp_regression.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409174/copp_regression.log)
- copp unit test log: 
[copp_unit_test.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409176/copp_unit_test.log)
- copp unit test script: 
[copp_unit_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409177/copp_unit_test.txt)

- dhcp_relay regression result: 
[dhcp_relay_regression-2024-02-20-17-42-04.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409184/dhcp_relay_regression-2024-02-20-17-42-04.html.pdf)
- dhcp_relay regression log: 
[dhcp_relay_regression.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409186/dhcp_relay_regression.log)
- dhcp_relay unit test log: 
[dhcp_relay_unit_test.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409191/dhcp_relay_unit_test.log)
- dhcp__relay unit test script: 
[dhcp_relay_unit_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409195/dhcp_relay_unit_test.txt)

- dhcp_snooping regression result: 
[dhcp_snooping_regression-2024-02-21-17-55-18.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409198/dhcp_snooping_regression-2024-02-21-17-55-18.html.pdf)
- dhcp_snooping regression log: 
[dhcp_snooping_regression.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409202/dhcp_snooping_regression.log)
- dhcp_snooping unit test log: 
[dhcp_snooping_unit_test.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409205/dhcp_snooping_unit_test.log)
- dhcp_snooping unit test script: 
[dhcp_snooping_unit_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14409210/dhcp_snooping_unit_test.txt)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Run regression test and lots of unit tests.

### CODE REVIEW

**Review by kerry-meyer — COMMENTED** [review]
The code changes and test results look good. I have posted one comment on a "hard coded" union of keys that could probably be calculated instead for better maintainability if the keys set(s) change in the future. Everything else looks fine.

File: plugins/module_utils/network/sonic/config/bfd/bfd.py [hunk 1]
```diff
@@ -43,6 +46,71 @@
...
+    'enabled': True,
+    'transmit_interval': 300,
+    'receive_interval': 300,
+    'detect_multiplier': 3,
+    'passive_mode': False,
+    'min_ttl': 254,
+    'echo_interval': 300,
+    'echo_mode': False
+}
+BFD_CONFIG_KEY_SET = {
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/bfd/bfd.py:59]: Thanks. 
They are  not all keys. They are keys with default values.

[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/bfd/bfd.py:59]: The preceding dictionary (BFD_CONFIG_DEFAULT) contains all keys with default values, but this one contains keys with no default value and it includes all keys that are present in any of the three types of BFD configuration lists.

[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/bfd/bfd.py]: Changed. Thanks!

**Review by kerry-meyer — APPROVED** [review]
Thanks for making the additional requested change and thank you for providing this enhancement of the check and diff functionality for our resource module collection.

All of the currently proposed code changes and test results look good.

Approved.

**Pull request merged**